### PR TITLE
Extended ceraos (ceragon) discovery to graph various wireless paramet…

### PR DESCRIPTION
**Pull request by barryodonovan**
#### Please note

> Please read this information carefully. You can run `./scripts/pre-commit.php` to check your code before submitting.

- [x] Have you signed the [Contributors agreement](http://docs.librenms.org/General/Contributing/)
- [x] Have you followed our [code guidelines?](http://docs.librenms.org/Developing/Code-Guidelines/)

…ers. This required adding a command for snmpgetnext also

### CODE REVIEW

File: tests/snmpsim/ceraos.snmprec [hunk 1]
```diff
@@ -1,2 +1 @@
-1.3.6.1.2.1.1.1.0|4|High capacity packet radio outdoor unit
-1.3.6.1.2.1.1.2.0|6|1.3.6.1.4.1.2281.1.20.2.2
+1.3.6.1.2.1.1.2.0|6|1.3.6.1.4.1.2281
```
Comment (laf) [inline tests/snmpsim/ceraos.snmprec]: This needs to stay as it was, the test is supposed to simulate a device and not the bare minimum the code is checking for.

File: includes/snmp.inc.php [hunk 2]
```diff
@@ -264,6 +282,29 @@ function snmp_get($device, $oid, $options = null, $mib = null, $mibdir = null)
 }//end snmp_get()
 
 
+function snmp_get_next($device, $oid, $options = null, $mib = null, $mibdir = null)
```
Comment (laf) [inline includes/snmp.inc.php]: What's the reason for adding snmp_get_next?

File: includes/snmp.inc.php [hunk 3]
```diff
@@ -264,6 +282,29 @@ function snmp_get($device, $oid, $options = null, $mib = null, $mibdir = null)
...
+{
+    global $runtime_stats;
+
+    if (strstr($oid, ' ')) {
+        echo report_this_text("snmp_get_next called for multiple OIDs: $oid");
```
Comment (laf) [inline includes/snmp.inc.php]: This should be d_echo(). We don't want to print things like this for normal poller/discovery runs.

File: includes/polling/os/ceraos.inc.php [hunk 4]
```diff
@@ -1,5 +1,63 @@
...
+ * Free Software Foundation, either version 3 of the License, or (at your
+ * option) any later version.  Please see LICENSE.txt at the top level of
+ * the source code distribution for details.
+ */
 
-$hardware = snmp_get($device, 'genEquipInventoryCardName.127', '-Osqnv', 'MWRM-UNIT-MIB');
-$serial = snmp_get($device, 'genEquipInventorySerialNumber.127', '-Osqnv', 'MWRM-UNIT-MIB');
-$version = snmp_get($device, 'genEquipMngSwIDUVersionsRunningVersion.1', '-Osqnv', 'MWRM-UNIT-MIB');
+$ceragon_type = snmp_get($device, 'sysObjectID.0', '-mSNMPv2-MIB -Oqv', '');
+if (strstr($ceragon_type, '.2281.1.10')) {
```
Comment (laf) [inline includes/polling/os/ceraos.inc.php]: This entire if/elseif block needs to be put into a rewrite function in `includes/rewrites.php` like `rewrite_junos_hardware()`

File: tests/snmpsim/ceraos.snmprec (hunk 1 above)
Comment (barryodonovan) [inline tests/snmpsim/ceraos.snmprec]: Ok, so it isn't used for live devices?

File: includes/snmp.inc.php (hunk 2 above)
Comment (barryodonovan) [inline includes/snmp.inc.php]: It is not possible to get the sub-oid of each subsequent radio without making assumptions.
snmp_get_next allows me to walk along the tree for the previously discovered radio count.

Also there are some oids which use index 0 for IP10 but index 127 for IP20C, they may use other indexes for other variants but each variant appears to only use one index so get_next will find it.

Ideally discovery should be provided for each major ( and possibly some minor varients ) eg a discovery for IP10, a different discovery for IP20C, E & S with possibly another one again for the N, etc variants as these provide multiple hot failover options which may not even be polled correctly.

That doesn't seem to fit the LibreNMS model though so I've tried to cover as many bases as possible.

Is it preferred to not use get_next?

File: includes/snmp.inc.php (hunk 3 above)
Comment (barryodonovan) [inline includes/snmp.inc.php]: OK

File: includes/polling/os/ceraos.inc.php (hunk 4 above)
Comment (barryodonovan) [inline includes/polling/os/ceraos.inc.php]: OK

**Comment by barryodonovan** [comment]
@laf Regarding Travis failure for php5.3, what versions do I need to support?

File: tests/snmpsim/ceraos.snmprec (hunk 1 above)
Comment (laf) [inline tests/snmpsim/ceraos.snmprec]: It's not no, it's used for the test units which should be as complete as possible. For instance someone could add an OS discovery based on sysName and match this device type before it got to this OS discovery. Having both sysObjectId and sysName in the test units means we catch more.

File: includes/snmp.inc.php (hunk 2 above)
Comment (laf) [inline includes/snmp.inc.php]: No preference to not use get_next it's just we haven't had the need for it before so we want to make sure we don't introduce functions that aren't needed.

I'm assuming you can't just bulkwalk?

**Comment by laf** [comment]
5.3 and above is still supported by us. The error is just one line so is an easy fix.

File: includes/snmp.inc.php (hunk 3 above)
Comment (barryodonovan) [inline includes/snmp.inc.php]: I've also changed this echo statement for the snmp_get function I based mine on.

File: includes/snmp.inc.php (hunk 2 above)
Comment (laf) [inline includes/snmp.inc.php]: Have you tried `snmpwalk -v2c -cCOMMUNITY HOSTNAME   -M /opt/librenms/mibs/ -m IF-MIB ifNumber`?

Comment (barryodonovan) [inline includes/snmp.inc.php]: I am only using the snmp_get_next to traverse the IF-MIB::ifIndex
Once I have my indexes I use the index to poll the various OIDs, eg genEquipRadioStatusDefectedBlocks.268451969, genEquipRadioStatusDefectedBlocks.268451970.

The other option I can see is like this:

if (IP10) {
    get genEquipRadioStatusDefectedBlocks.1
} else if (IP20C) {
    get genEquipRadioStatusDefectedBlocks.268451969
    get genEquipRadioStatusDefectedBlocks.268451970
} elseif (IP20S) {
    get genEquipRadioStatusDefectedBlocks.?????
} elseif (IP20N) {
    get genEquipRadioStatusDefectedBlocks.?????
    get genEquipRadioStatusDefectedBlocks.?????
    get genEquipRadioStatusDefectedBlocks.?????
    get genEquipRadioStatusDefectedBlocks.?????
} elseif ...


I'm trying to make the discovery and polling as generic (and extensible) as possible for various variants of ceragon radios since I don't have all of them to test against.

Comment (laf) [inline includes/snmp.inc.php]: mib polling afaik should just walk through genEquipRadioStatusDefectedBlocks and pick out the data it finds. Have you tried just specifying that?

Comment (barryodonovan) [inline includes/snmp.inc.php]: I'm not sure I see the benefit of this? 
I still need to count the number of radio interfaces to make the information available for graphing, I will also need to sort the result of each snmp_walk into the appropriate array for mib polling and rrd ds insertion.

I will have to do lots of hoop-jumping and write at least a few very device specific functions in order to avoid the creation and use of one very generic function which exposes a basic NET-SNMP (snmpgetnext) function to LibreNMS.

**Comment by barryodonovan** [comment]
@laf Are you happy for this to be merged? I'd really like to be able to have it in production.

**Comment by barryodonovan** [comment]
@laf @paulgear I have removed all use of snmp_get_next. Can this be merged now?

**Comment by barryodonovan** [comment]
@laf @paulgear 

Bump!

**Comment by laf** [comment]
Are you able to provide the output of:

`snmpbulkwalk -OUneb -v2c -c COMMUNITY HOSTNAME .`

It would help us test this and see what we can do to get it merged in. At the moment this could lead to a lot of snmp queries to just get client count never mind the wifi polling and we've spent quite a bit of time trying to optimise polling in other areas.

**Comment by barryodonovan** [comment]
See below.


.1.3.6.1.2.1.1.1.0 = STRING: "High capacity packet radio outdoor unit"
.1.3.6.1.2.1.1.2.0 = OID: .1.3.6.1.4.1.2281.1.20.2.2
.1.3.6.1.2.1.1.3.0 = INTEGER: ***s/n***
.1.3.6.1.2.1.1.4.0 = ""
.1.3.6.1.2.1.1.5.0 = STRING: "Hostname"
.1.3.6.1.2.1.1.6.0 = STRING: "Location"
.1.3.6.1.2.1.2.1.0 = INTEGER: 5
.1.3.6.1.2.1.2.2.1.1.268443713 = INTEGER: 268443713
.1.3.6.1.2.1.2.2.1.1.268443714 = INTEGER: 268443714
.1.3.6.1.2.1.2.2.1.1.268451969 = INTEGER: 268451969
.1.3.6.1.2.1.2.2.1.1.268451970 = INTEGER: 268451970
.1.3.6.1.2.1.2.2.1.1.268562561 = INTEGER: 268562561
.1.3.6.1.2.1.2.2.1.2.268443713 = STRING: "Ethernet"
.1.3.6.1.2.1.2.2.1.2.268443714 = STRING: "Ethernet"
.1.3.6.1.2.1.2.2.1.2.268451969 = STRING: "Radio"
.1.3.6.1.2.1.2.2.1.2.268451970 = STRING: "Radio"
.1.3.6.1.2.1.2.2.1.2.268562561 = STRING: "Multi-Carrier ABC"
.1.3.6.1.2.1.2.2.1.3.268443713 = INTEGER: 6
.1.3.6.1.2.1.2.2.1.3.268443714 = INTEGER: 6
.1.3.6.1.2.1.2.2.1.3.268451969 = INTEGER: 1
.1.3.6.1.2.1.2.2.1.3.268451970 = INTEGER: 1
.1.3.6.1.2.1.2.2.1.3.268562561 = INTEGER: 1
.1.3.6.1.2.1.2.2.1.4.268443713 = INTEGER: 1632
.1.3.6.1.2.1.2.2.1.4.268443714 = INTEGER: 1632
.1.3.6.1.2.1.2.2.1.4.268451969 = INTEGER: 2000
.1.3.6.1.2.1.2.2.1.4.268451970 = INTEGER: 2000
.1.3.6.1.2.1.2.2.1.4.268562561 = INTEGER: 2000
.1.3.6.1.2.1.2.2.1.5.268443713 = INTEGER: 1000000000
.1.3.6.1.2.1.2.2.1.5.268443714 = INTEGER: 10000000
.1.3.6.1.2.1.2.2.1.5.268451969 = INTEGER: 349341000
.1.3.6.1.2.1.2.2.1.5.268451970 = INTEGER: 349341000
.1.3.6.1.2.1.2.2.1.5.268562561 = INTEGER: 1000000000
.1.3.6.1.2.1.2.2.1.6.268443713 = Hex-STRING: 00 00 00 00 00 00 
.1.3.6.1.2.1.2.2.1.6.268443714 = Hex-STRING: 00 00 00 00 00 00 
.1.3.6.1.2.1.2.2.1.6.268451969 = Hex-STRING: 00 00 00 00 00 00 
.1.3.6.1.2.1.2.2.1.6.268451970 = Hex-STRING: 00 00 00 00 00 00 
.1.3.6.1.2.1.2.2.1.6.268562561 = Hex-STRING: 00 0A 25 63 6B 79 
.1.3.6.1.2.1.2.2.1.7.268443713 = INTEGER: 1
.1.3.6.1.2.1.2.2.1.7.268443714 = INTEGER: 2
.1.3.6.1.2.1.2.2.1.7.268451969 = INTEGER: 1
.1.3.6.1.2.1.2.2.1.7.268451970 = INTEGER: 1
.1.3.6.1.2.1.2.2.1.7.268562561 = INTEGER: 1
.1.3.6.1.2.1.2.2.1.8.268443713 = INTEGER: 1
.1.3.6.1.2.1.2.2.1.8.268443714 = INTEGER: 2
.1.3.6.1.2.1.2.2.1.8.268451969 = INTEGER: 2
.1.3.6.1.2.1.2.2.1.8.268451970 = INTEGER: 2
.1.3.6.1.2.1.2.2.1.8.268562561 = INTEGER: 2
.1.3.6.1.2.1.2.2.1.9.268443713 = INTEGER: 1478133526
.1.3.6.1.2.1.2.2.1.9.268443714 = INTEGER: 1
.1.3.6.1.2.1.2.2.1.9.268451969 = INTEGER: 1479966796
.1.3.6.1.2.1.2.2.1.9.268451970 = INTEGER: 1479966796
.1.3.6.1.2.1.2.2.1.9.268562561 = INTEGER: 1479966796
.1.3.6.1.2.1.2.2.1.10.268443713 = Counter32: 2369600066
.1.3.6.1.2.1.2.2.1.10.268443714 = Counter32: 0
.1.3.6.1.2.1.2.2.1.10.268451969 = Counter32: 0
.1.3.6.1.2.1.2.2.1.10.268451970 = Counter32: 0
.1.3.6.1.2.1.2.2.1.10.268562561 = Counter32: 2592959683
.1.3.6.1.2.1.2.2.1.11.268443713 = Counter32: 4247867654
.1.3.6.1.2.1.2.2.1.11.268443714 = Counter32: 0
.1.3.6.1.2.1.2.2.1.11.268451969 = Counter32: 0
.1.3.6.1.2.1.2.2.1.11.268451970 = Counter32: 0
.1.3.6.1.2.1.2.2.1.11.268562561 = Counter32: 891805690
.1.3.6.1.2.1.2.2.1.13.268443713 = Counter32: 0
.1.3.6.1.2.1.2.2.1.13.268443714 = Counter32: 0
.1.3.6.1.2.1.2.2.1.13.268451969 = Counter32: 0
.1.3.6.1.2.1.2.2.1.13.268451970 = Counter32: 0
.1.3.6.1.2.1.2.2.1.13.268562561 = Counter32: 0
.1.3.6.1.2.1.2.2.1.14.268443713 = Counter32: 0
.1.3.6.1.2.1.2.2.1.14.268443714 = Counter32: 0
.1.3.6.1.2.1.2.2.1.14.268451969 = Counter32: 0
.1.3.6.1.2.1.2.2.1.14.268451970 = Counter32: 0
.1.3.6.1.2.1.2.2.1.14.268562561 = Counter32: 13
.1.3.6.1.2.1.2.2.1.16.268443713 = Counter32: 2605250159
.1.3.6.1.2.1.2.2.1.16.268443714 = Counter32: 12309578
.1.3.6.1.2.1.2.2.1.16.268451969 = Counter32: 0
.1.3.6.1.2.1.2.2.1.16.268451970 = Counter32: 0
.1.3.6.1.2.1.2.2.1.16.268562561 = Counter32: 2362088934
.1.3.6.1.2.1.2.2.1.17.268443713 = Counter32: 891805677
.1.3.6.1.2.1.2.2.1.17.268443714 = Counter32: 0
.1.3.6.1.2.1.2.2.1.17.268451969 = Counter32: 0
.1.3.6.1.2.1.2.2.1.17.268451970 = Counter32: 0
.1.3.6.1.2.1.2.2.1.17.268562561 = Counter32: 4250227913
.1.3.6.1.2.1.2.2.1.19.268443713 = Counter32: 0
.1.3.6.1.2.1.2.2.1.19.268443714 = Counter32: 0
.1.3.6.1.2.1.2.2.1.19.268451969 = Counter32: 0
.1.3.6.1.2.1.2.2.1.19.268451970 = Counter32: 0
.1.3.6.1.2.1.2.2.1.19.268562561 = Counter32: 0
.1.3.6.1.2.1.2.2.1.20.268443713 = Counter32: 0
.1.3.6.1.2.1.2.2.1.20.268443714 = Counter32: 0
.1.3.6.1.2.1.2.2.1.20.268451969 = Counter32: 0
.1.3.6.1.2.1.2.2.1.20.268451970 = Counter32: 0
.1.3.6.1.2.1.2.2.1.20.268562561 = Counter32: 0
.1.3.6.1.2.1.16.1.1.1.1.268443713 = INTEGER: 268443713
.1.3.6.1.2.1.16.1.1.1.1.268443714 = INTEGER: 268443714
.1.3.6.1.2.1.16.1.1.1.1.268562561 = INTEGER: 268562561
.1.3.6.1.2.1.16.1.1.1.4.268443713 = Counter32: 2369600148
.1.3.6.1.2.1.16.1.1.1.4.268443714 = Counter32: 0
.1.3.6.1.2.1.16.1.1.1.4.268562561 = Counter32: 2592959683
.1.3.6.1.2.1.16.1.1.1.5.268443713 = Counter32: 4250290150
.1.3.6.1.2.1.16.1.1.1.5.268443714 = Counter32: 0
.1.3.6.1.2.1.16.1.1.1.5.268562561 = Counter32: 893625196
.1.3.6.1.2.1.16.1.1.1.6.268443713 = Counter32: 38023
.1.3.6.1.2.1.16.1.1.1.6.268443714 = Counter32: 0
.1.3.6.1.2.1.16.1.1.1.6.268562561 = Counter32: 30603
.1.3.6.1.2.1.16.1.1.1.7.268443713 = Counter32: 2384473
.1.3.6.1.2.1.16.1.1.1.7.268443714 = Counter32: 0
.1.3.6.1.2.1.16.1.1.1.7.268562561 = Counter32: 1788903
.1.3.6.1.2.1.16.1.1.1.8.268443713 = Counter32: 0
.1.3.6.1.2.1.16.1.1.1.8.268443714 = Counter32: 0
.1.3.6.1.2.1.16.1.1.1.8.268562561 = Counter32: 13
.1.3.6.1.2.1.16.1.1.1.9.268443713 = Counter32: 0
.1.3.6.1.2.1.16.1.1.1.9.268443714 = Counter32: 0
.1.3.6.1.2.1.16.1.1.1.9.268562561 = Counter32: 0
.1.3.6.1.2.1.16.1.1.1.10.268443713 = Counter32: 0
.1.3.6.1.2.1.16.1.1.1.10.268443714 = Counter32: 0
.1.3.6.1.2.1.16.1.1.1.10.268562561 = Counter32: 0
.1.3.6.1.2.1.16.1.1.1.11.268443713 = Counter32: 0
.1.3.6.1.2.1.16.1.1.1.11.268443714 = Counter32: 0
.1.3.6.1.2.1.16.1.1.1.11.268562561 = Counter32: 0
.1.3.6.1.2.1.16.1.1.1.12.268443713 = Counter32: 0
.1.3.6.1.2.1.16.1.1.1.12.268443714 = Counter32: 0
.1.3.6.1.2.1.16.1.1.1.12.268562561 = Counter32: 0
.1.3.6.1.2.1.16.1.1.1.14.268443713 = Counter32: 2881812152
.1.3.6.1.2.1.16.1.1.1.14.268443714 = Counter32: 0
.1.3.6.1.2.1.16.1.1.1.14.268562561 = Counter32: 651579336
.1.3.6.1.2.1.16.1.1.1.15.268443713 = Counter32: 2858410072
.1.3.6.1.2.1.16.1.1.1.15.268443714 = Counter32: 0
.1.3.6.1.2.1.16.1.1.1.15.268562561 = Counter32: 1706668412
.1.3.6.1.2.1.16.1.1.1.16.268443713 = Counter32: 588484226
.1.3.6.1.2.1.16.1.1.1.16.268443714 = Counter32: 0
.1.3.6.1.2.1.16.1.1.1.16.268562561 = Counter32: 539059226
.1.3.6.1.2.1.16.1.1.1.17.268443713 = Counter32: 348724709
.1.3.6.1.2.1.16.1.1.1.17.268443714 = Counter32: 0
.1.3.6.1.2.1.16.1.1.1.17.268562561 = Counter32: 362722511
.1.3.6.1.2.1.16.1.1.1.18.268443713 = Counter32: 454337576
.1.3.6.1.2.1.16.1.1.1.18.268443714 = Counter32: 0
.1.3.6.1.2.1.16.1.1.1.18.268562561 = Counter32: 527352206
.1.3.6.1.2.1.16.1.1.1.19.268443713 = Counter32: 1413488711
.1.3.6.1.2.1.16.1.1.1.19.268443714 = Counter32: 0
.1.3.6.1.2.1.16.1.1.1.19.268562561 = Counter32: 1401210801
.1.3.6.1.2.1.16.1.7.1.4.268443713 = Counter64: 3597257226982
.1.3.6.1.2.1.16.1.7.1.4.268443714 = Counter64: 0
.1.3.6.1.2.1.16.1.7.1.4.268562561 = Counter64: 28233412996291
.1.3.6.1.2.1.31.1.1.1.1.268443713 = STRING: "Ethernet: Slot 1, port 1"
.1.3.6.1.2.1.31.1.1.1.1.268443714 = STRING: "Ethernet: Slot 1, port 2"
.1.3.6.1.2.1.31.1.1.1.1.268451969 = STRING: "Radio: Slot 2, port 1"
.1.3.6.1.2.1.31.1.1.1.1.268451970 = STRING: "Radio: Slot 2, port 2"
.1.3.6.1.2.1.31.1.1.1.1.268562561 = STRING: "Multi Carrier ABC: Group #1"
.1.3.6.1.2.1.31.1.1.1.2.268443713 = Counter32: 2384474
.1.3.6.1.2.1.31.1.1.1.2.268443714 = Counter32: 0
.1.3.6.1.2.1.31.1.1.1.2.268451969 = Counter32: 0
.1.3.6.1.2.1.31.1.1.1.2.268451970 = Counter32: 0
.1.3.6.1.2.1.31.1.1.1.2.268562561 = Counter32: 1788903
.1.3.6.1.2.1.31.1.1.1.3.268443713 = Counter32: 38023
.1.3.6.1.2.1.31.1.1.1.3.268443714 = Counter32: 0
.1.3.6.1.2.1.31.1.1.1.3.268451969 = Counter32: 0
.1.3.6.1.2.1.31.1.1.1.3.268451970 = Counter32: 0
.1.3.6.1.2.1.31.1.1.1.3.268562561 = Counter32: 30603
.1.3.6.1.2.1.31.1.1.1.4.268443713 = Counter32: 1864817
.1.3.6.1.2.1.31.1.1.1.4.268443714 = Counter32: 75914
.1.3.6.1.2.1.31.1.1.1.4.268451969 = Counter32: 0
.1.3.6.1.2.1.31.1.1.1.4.268451970 = Counter32: 0
.1.3.6.1.2.1.31.1.1.1.4.268562561 = Counter32: 0
.1.3.6.1.2.1.31.1.1.1.5.268443713 = Counter32: 144633
.1.3.6.1.2.1.31.1.1.1.5.268443714 = Counter32: 114030
.1.3.6.1.2.1.31.1.1.1.5.268451969 = Counter32: 0
.1.3.6.1.2.1.31.1.1.1.5.268451970 = Counter32: 0
.1.3.6.1.2.1.31.1.1.1.5.268562561 = Counter32: 0
.1.3.6.1.2.1.31.1.1.1.6.268443713 = Counter64: 3597257226982
.1.3.6.1.2.1.31.1.1.1.6.268443714 = Counter64: 0
.1.3.6.1.2.1.31.1.1.1.6.268451969 = Counter64: 0
.1.3.6.1.2.1.31.1.1.1.6.268451970 = Counter64: 0
.1.3.6.1.2.1.31.1.1.1.6.268562561 = Counter64: 28233412996291
.1.3.6.1.2.1.31.1.1.1.10.268443713 = Counter64: 28233425286767
.1.3.6.1.2.1.31.1.1.1.10.268443714 = Counter64: 12309578
.1.3.6.1.2.1.31.1.1.1.10.268451969 = Counter64: 0
.1.3.6.1.2.1.31.1.1.1.10.268451970 = Counter64: 0
.1.3.6.1.2.1.31.1.1.1.10.268562561 = Counter64: 3678854094310
.1.3.6.1.2.1.31.1.1.1.14.268443713 = INTEGER: 1
.1.3.6.1.2.1.31.1.1.1.14.268443714 = INTEGER: 1
.1.3.6.1.2.1.31.1.1.1.14.268451969 = INTEGER: 1
.1.3.6.1.2.1.31.1.1.1.14.268451970 = INTEGER: 1
.1.3.6.1.2.1.31.1.1.1.14.268562561 = INTEGER: 1
.1.3.6.1.2.1.31.1.1.1.15.268443713 = INTEGER: 1000000000
.1.3.6.1.2.1.31.1.1.1.15.268443714 = INTEGER: 10000000
.1.3.6.1.2.1.31.1.1.1.15.268451969 = INTEGER: 349341000
.1.3.6.1.2.1.31.1.1.1.15.268451970 = INTEGER: 349341000
.1.3.6.1.2.1.31.1.1.1.15.268562561 = INTEGER: 1000000000
.1.3.6.1.2.1.31.1.1.1.17.268443713 = INTEGER: 2
.1.3.6.1.2.1.31.1.1.1.17.268443714 = INTEGER: 2
.1.3.6.1.2.1.31.1.1.1.17.268451969 = INTEGER: 2
.1.3.6.1.2.1.31.1.1.1.17.268451970 = INTEGER: 2
.1.3.6.1.2.1.31.1.1.1.17.268562561 = INTEGER: 2
.1.3.6.1.4.1.2281.10.1.1.1.0 = INTEGER: 48
.1.3.6.1.4.1.2281.10.1.1.8.0 = INTEGER: 0
.1.3.6.1.4.1.2281.10.1.1.9.0 = INTEGER: 40
.1.3.6.1.4.1.2281.10.1.1.10.0 = INTEGER: 44
.1.3.6.1.4.1.2281.10.1.1.11.6.6.1.1.1 = INTEGER: 1
.1.3.6.1.4.1.2281.10.1.1.11.6.6.1.2.1 = INTEGER: 1024
.1.3.6.1.4.1.2281.10.1.1.11.6.6.1.3.1 = IpAddress: 10.1.2.3


It just times out after this.

**Comment by barryodonovan** [comment]
@laf  I've pasted the results of a polling run against a ceragon in http://pastebin.com/P5Z6F259

**Comment by barryodonovan** [comment]
@f0o @paulgear @SaaldjorMike 
Would anyone be able to look at this PR to merge it? I'd like to get it into production and move on with supporting some of the other hardware we use.

**Comment by Rosiak** [comment]
From a code perspective it looks ok, could be a bit picky with variable naming, but in this case I won't bother. So I'm good with this, does not break anything for me.

File: includes/polling/os/ceraos.inc.php [hunk 5]
```diff
@@ -1,5 +1,31 @@
...
+ * under the terms of the GNU General Public License as published by the
+ * Free Software Foundation, either version 3 of the License, or (at your
+ * option) any later version.  Please see LICENSE.txt at the top level of
+ * the source code distribution for details.
+ */
 
-$hardware = snmp_get($device, 'genEquipInventoryCardName.127', '-Osqnv', 'MWRM-UNIT-MIB');
-$serial = snmp_get($device, 'genEquipInventorySerialNumber.127', '-Osqnv', 'MWRM-UNIT-MIB');
-$version = snmp_get($device, 'genEquipMngSwIDUVersionsRunningVersion.1', '-Osqnv', 'MWRM-UNIT-MIB');
+$ceragon_type = snmp_get($device, 'sysObjectID.0', '-mSNMPv2-MIB -Oqv', '');
```
Comment (laf) [inline includes/polling/os/ceraos.inc.php]: sysObjectId is already available as `$poll_device['sysObjectID']`. The snmp_get in general isn't formatted correctly though. you don't need to add -mMIBNAME. See exampels elsewhere in the code.

File: includes/polling/os/ceraos.inc.php [hunk 6]
```diff
@@ -1,5 +1,31 @@
...
+$hardware = rewrite_ceraos_hardware($ceragon_type); // function in ./includes/rewrites.php
+$version = snmp_get($device, 'genEquipMngSwIDUVersionsRunningVersion.1', '-mMWRM-RADIO-MIB -Oqv', '');
```
Comment (laf) [inline includes/polling/os/ceraos.inc.php]: Same here for -mMIBNAME

File: includes/polling/os/ceraos.inc.php [hunk 7]
```diff
@@ -1,5 +1,31 @@
...
+$serial = snmp_walk($device, 'genEquipInventorySerialNumber', '-mMWRM-RADIO-MIB -Oqv', '');
```
Comment (laf) [inline includes/polling/os/ceraos.inc.php]: Same for the next three for -mMIBNAME

Also, why snmpwalk for a single item, surely snmpget is better? Or really multi get for the 4 MWRM-RADIO-MIB you want.

File: includes/polling/os/ceraos.inc.php [hunk 8]
```diff
@@ -1,5 +1,31 @@
...
+$latitude = snmp_walk($device, 'genEquipUnitLatitude', '-mMWRM-RADIO-MIB -Oqv', '');
+$longitude = snmp_walk($device, 'genEquipUnitLongitude', '-mMWRM-RADIO-MIB -Oqv', '');
+
+$IfIndex = 0;
+$ifIndexesArray = array();
+$ifIndexesArray = explode("\n", snmp_walk($device, "ifIndex", "-mIF-MIB -Oqv", ""));
```
Comment (laf) [inline includes/polling/os/ceraos.inc.php]: I'm not a fan of walking the ifIndex table in os polling. Is there absolutely no other way this can be done.

File: includes/polling/os/ceraos.inc.php [hunk 9]
```diff
@@ -1,5 +1,31 @@
...
+$num_radios = 0;
+
+foreach ($ifIndexesArray as $IfIndex) {
+    $IfDescr = snmp_get($device, "ifDescr.$IfIndex", "-mIF-MIB -Oqv", "");
```
Comment (laf) [inline includes/polling/os/ceraos.inc.php]: Multi get would be better here if the walk has to stay.

File: includes/polling/os/ceraos.inc.php [hunk 10]
```diff
@@ -1,5 +1,31 @@
...
+    $IfName = snmp_get($device, "ifName.$IfIndex", "-mIF-MIB -Oqv", "");
+    if (stristr($IfDescr, "Radio")) {
+        $num_radios = $num_radios+1;
+    }
+}
+$features = $num_radios . " radios in unit";
```
Comment (laf) [inline includes/polling/os/ceraos.inc.php]: In fact we are just using the `$num_radios` to put in the feature. Doing all of that is expensive just to add the number of radios to a descriptive field.

File: includes/polling/mib/ceraos-mib.inc.php [hunk 11]
```diff
@@ -0,0 +1,86 @@
...
+
+$features = explode(' ', $device[features]);
+$num_radios = $features[0];
+
+$mib_oids = array();
+
+$radioNumber = 0;
+$IfIndex = 0;
+$ifIndexesArray = array();
+$ifIndexesArray = explode("\n", snmp_walk($device, "ifIndex", "-mIF-MIB -Oqv", ""));
```
Comment (laf) [inline includes/polling/mib/ceraos-mib.inc.php]: -mMIBNAME again in this file. They all need to be updated properly.

Comment (laf) [inline includes/polling/mib/ceraos-mib.inc.php]: Can we not do something a bit better here where we pass the ifEntry cache from ports polling module to this module as we already have this info and I think polling modules are run in alpha numeric order.

File: includes/polling/os/ceraos.inc.php (hunk 5 above)
Comment (barryodonovan) [inline includes/polling/os/ceraos.inc.php]: If I don't use -mMIB-NAME, I get the following:
genEquipInventorySerialNumber.0: Unknown Object Identifier (Sub-id not found: (top) -> genEquipInventorySerialNumber)
I know it is unlikely that the numerical OID would change in eg IF-MIB but it is possible since the textual OID is canonical and I think it is far more possible that it will change in the ceragon MIBs in order to support future variants.

File: includes/polling/os/ceraos.inc.php (hunk 7 above)
Comment (barryodonovan) [inline includes/polling/os/ceraos.inc.php]: Just to point out that snmp_get_multi() is a disaster!

It was obviously written for a specific use case and does not easily support any other use.
I am getting $version, $lat, $long as multi_get. $sersion is index 1, lat and log are index 0. result is:

[1] => [OID for version] = version
[0] => [OID for lat] = lat
          [OID for long] = long

and all because someone decided to munge their data in a generic function...

File: includes/polling/os/ceraos.inc.php (hunk 5 above)
Comment (laf) [inline includes/polling/os/ceraos.inc.php]: You can call snmp_get though with the MIB name already, like so:

```
$ceragon_type = snmp_get($device, 'sysObjectID.0', '-Oqv', 'SNMPv2-MIB');
```

File: includes/polling/os/ceraos.inc.php (hunk 7 above)
Comment (barryodonovan) [inline includes/polling/os/ceraos.inc.php]: Now I'm going to clean up the result I get back out of it which will make the function even harder to clean up in future, and I'm sure I'm not the first...

File: includes/polling/os/ceraos.inc.php (hunk 8 above)
Comment (barryodonovan) [inline includes/polling/os/ceraos.inc.php]: snmp_get_next?

File: includes/polling/os/ceraos.inc.php (hunk 10 above)
Comment (barryodonovan) [inline includes/polling/os/ceraos.inc.php]: I need to know the number of radios in order to determine how many lines are in each graph. Otherwise I do it for each graph individually, plus for for my mib poll...

File: includes/polling/mib/ceraos-mib.inc.php (hunk 11 above)
Comment (barryodonovan) [inline includes/polling/mib/ceraos-mib.inc.php]: ports module unsets the variable...

File: includes/polling/os/ceraos.inc.php (hunk 8 above)
Comment (laf) [inline includes/polling/os/ceraos.inc.php]: lol, I'm on irc at the moment if you're around.

File: includes/polling/os/ceraos.inc.php (hunk 7 above)
Comment (barryodonovan) [inline includes/polling/os/ceraos.inc.php]: Array
(
    [1] => Array
        (
            [MWRM-UNIT-MIB::genEquipMngSwIDUVersionsRunningVersion] => STRING: "8.0.0.0.0.338"
        )

    [0] => Array
        (
            [MWRM-UNIT-MIB::genEquipUnitLatitude] => STRING: "52.968708"
            [MWRM-UNIT-MIB::genEquipUnitLongitude] => STRING: "-8.718439"
        )

)

Comment (laf) [inline includes/polling/os/ceraos.inc.php]: Can you get the raw snmp output for this multi get call please?

File: includes/definitions.inc.php [hunk 12]
```diff
@@ -766,7 +766,11 @@
 $config['os'][$os]['icon']             = 'ceragon';
 $config['os'][$os]['over'][0]['graph'] = 'device_bits';
 $config['os'][$os]['over'][0]['text']  = 'Traffic';
-
+$config['os'][$os]['over'][1]['graph'] = 'device_ceraos_RxLevel';
+$config['os'][$os]['over'][1]['text']  = 'Rx Level';
+$config['os'][$os]['over'][2]['graph'] = 'device_ping_perf';
+$config['os'][$os]['over'][2]['text']  = 'Ping Times';
+$config['os'][$os]['mib-dir'][]        = 'ceraos';
```
Comment (laf) [inline includes/definitions.inc.php]: It would be mib_dir but this isn't needed anyway as a directory of the same os name is checked anyway so you can just remove this line.

File: includes/snmp.inc.php [hunk 13]
```diff
@@ -247,7 +247,7 @@ function snmp_get($device, $oid, $options = null, $mib = null, $mibdir = null)
     $time_start = microtime(true);
 
     if (strstr($oid, ' ')) {
-        echo report_this_text("snmp_get called for multiple OIDs: $oid");
+        echo(report_this_text("snmp_get called for multiple OIDs: $oid"));
```
Comment (laf) [inline includes/snmp.inc.php]: () around echo isn't needed.

**Comment by laf** [comment]
Just two comments then good for a merge.
